### PR TITLE
Allow public instance methods

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by shargon** [comment]
I think instance methods can create confusion with how ethereum works. They preserves instance states for future executions automatically (variables), and we don't (require to call storage), I don't know if allowing instance methods will cause developer confusion.

**Comment by erikzhang** [comment]
We need instance methods to support Inheritance. For example:

```c#
public class TokenContract : SmartContract
{
    public abstract string Symbol { get; }
    public abstract byte Decimals { get; }
}

public sealed class Nep17Token : TokenContract
{
    public override string Symbol => "MyToken";
    public override byte Decimals => 9;
}
```

**Comment by shargon** [comment]
Maybe we can deny instance variables in order to avoid confusion?

**Comment by erikzhang** [comment]
How about we call the constructor automatically?

**Comment by erikzhang** [comment]
> How about we call the constructor automatically?

Done.

File: src/Neo.Compiler.CSharp/MethodConvert.cs [hunk 1]
```diff
@@ -140,6 +140,29 @@ public void Convert(SemanticModel model)
...
+            {
+                for (int i = fields.Length - 1; i >= 0; i--)
+                    PushDefault(fields[i].Type);
+                Push(fields.Length);
+                AddInstruction(OpCode.PACK);
+            }
+            IMethodSymbol? constructor = type.InstanceConstructors.FirstOrDefault(p => p.Parameters.Length == 0);
+            if (constructor is null)
+                throw new CompilationException(type, DiagnosticId.NoParameterlessConstructor, "The contract class requires a parameterless constructor.");
+            Call(model, constructor, true, Array.Empty<ArgumentSyntax>());
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert.cs:161]: The constructor will be called only once if we call twice the instance method?

Comment (erikzhang) [reply src/Neo.Compiler.CSharp/MethodConvert.cs:161]: The forward method will call the constructor automatically. The real instance methods are not exposed.

**Review by shargon — APPROVED** [review]

**Pull request merged**